src/timestamp.js

```diff
@@ -0,0 +1,152 @@
+  }
+
+  /**
+   * Creates a new timestamp from the given date.
+   *
+   * @param {Date} date The date to initialize the `Timestamp` from.
+   * @return {Timestamp} A new `Timestamp` representing the same point in time
+   * as the given date.
+   */
+  static fromDate(date) {
+    return Timestamp.fromMillis(date.getTime());
+  }
+
+  /**
+   * Creates a new timestamp from the given number of milliseconds.
+   *
+   * @param {number} milliseconds Number of milliseconds since Unix epoch
+   * 1970-01-01T00:00:00Z.
+   * @return {Timestamp}  A new `Timestamp` representing the same point in time
+   * as the given number of milliseconds.
+   */
+  static fromMillis(milliseconds) {
+    const seconds = Math.floor(milliseconds / 1000);
+    const nanos = (milliseconds - seconds * 1000) * MS_TO_NANOS;
+    return new Timestamp(seconds, nanos);
+  }
+
+  /**
+   * Creates a new timestamp.
+   *
+   * @param {number} seconds The number of seconds of UTC time since Unix epoch
+   * 1970-01-01T00:00:00Z. Must be from 0001-01-01T00:00:00Z to
+   * 9999-12-31T23:59:59Z inclusive.
+   * @param {number} nanoseconds The non-negative fractions of a second at
+   * nanosecond resolution. Negative second values with fractions must still
+   * have non-negative nanoseconds values that count forward in time. Must be
+   * from 0 to 999,999,999 inclusive.
+   */
+  constructor(seconds, nanoseconds) {
+    validate.isInteger('seconds', seconds);
+    validate.isInteger('nanoseconds', nanoseconds, 0, 999999999);
+
+    this._seconds = seconds;
+    this._nanoseconds = nanoseconds;
+  }
+
+  /**
+   * The number of seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z.
+   *
+   * @type {number}
+   */
+  get seconds() {
+    return this._seconds;
+  }
+
+  /**
+   * The non-negative fractions of a second at nanosecond resolution.
+   *
+   * @type {number}
+   */
+  get nanoseconds() {
+    return this._nanoseconds;
+  }
+
+  /**
+   * Returns a new `Date` corresponding to this timestamp. This may lose
+   * precision.
+   *
+   * @return {Date} JavaScript `Date` object representing the same point in time
+   * as this `Timestamp`, with millisecond precision.
+   */
+  toDate() {
+    return new Date(
+      this._seconds * 1000 + Math.round(this._nanoseconds / MS_TO_NANOS)
+    );
+  }
+
+  /**
+   * Returns the number of milliseconds since Unix epoch 1970-01-01T00:00:00Z.
+   *
+   * @return {number} The point in time corresponding to this timestamp,
+   * represented as the number of milliseconds since Unix epoch
+   * 1970-01-01T00:00:00Z.
+   */
+  toMillis() {
+    return this._seconds * 1000 + Math.floor(this._nanoseconds / MS_TO_NANOS);
+  }
+
+  /**
+   * Returns 'true' if this `Timestamp` is equal to the provided one.
+   *
+   * @param {any} other The `Timestamp` to compare against.
+   * @return {boolean} 'true' if this `Timestamp` is equal to the provided one.
+   */
+  isEqual(other) {
+    return (
+      this === other ||
```

Comment:
Question: is there a case where the first check (`this === other`) succeeds but the second check doesn't? If not, is the first part necessary?